src/v/storage/segment.cc

```diff
@@ -448,6 +449,10 @@ ss::future<append_result> segment::do_append(const model::record_batch& b) {
                                            && has_compaction_index();
           const bool has_error = append_fut.failed() || index_append_failed;
           if (!has_error) {
+              if (!this->_first_write.has_value()) {
+                  // record time of first write
+                  this->_first_write = std::chrono::system_clock::now();
```

Comment:
This is a good question. At first, data batches makes sense if the goal of the config is to ensure that Kafka data can be deleted in a timely manner. That said, there's a case to be made for all batch types if we want this to be a tool to reduce disk space when the node isn't very inactive (though I suspect non-data-batches won't take much space at all).

Regardless, we should definitely be careful around `record_batch_type::raft_configuration` since that gets written when there's a leadership change, which also includes a segment roll.

Reply:
Leadership changes were what I had in mind when writing the question. If the segment rolls, and we immediately write a config batch, then the semantics of `segment.ms` are unrelated to user data (assuming we use the first batch of any type). It depends on the semantics that we want. To me, starting the timer after the first user data batch hits the partition makes most sense.

Reply:
agreed to limit the scope to data. I added model::record_batch_type::raft_data, should i use a different type?

Reply:
I don't think so. `raft_data` are the data batches.